script/tool/test/pubspec_check_command_test.dart

```diff
@@ -317,12 +404,162 @@ ${flutterSection(isPlugin: true)}
 ${dependenciesSection()}
 ''');
 
-      final Future<List<String>> result =
-          runCapturingPrint(runner, <String>['pubspec-check']);
+      Error? commandError;
+      final List<String> output = await runCapturingPrint(
+          runner, <String>['pubspec-check'], errorHandler: (Error e) {
+        commandError = e;
+      });
+
+      expect(commandError, isA<ToolExit>());
+      expect(
+        output,
+        containsAllInOrder(<Matcher>[
+          contains(
+              'Major sections should follow standard repository ordering:'),
+        ]),
+      );
+    });
+
+    test('fails when an implemenation package is missing "implements"',
+        () async {
+      final Directory pluginDirectory = createFakePlugin(
+          'plugin_a_foo', packagesDir.childDirectory('plugin_a'));
+
+      pluginDirectory.childFile('pubspec.yaml').writeAsStringSync('''
+${headerSection('plugin_a_foo', isPlugin: true)}
+${environmentSection()}
+${flutterSection(isPlugin: true)}
+${dependenciesSection()}
+${devDependenciesSection()}
+''');
+
+      Error? commandError;
+      final List<String> output = await runCapturingPrint(
+          runner, <String>['pubspec-check'], errorHandler: (Error e) {
+        commandError = e;
+      });
+
+      expect(commandError, isA<ToolExit>());
+      expect(
+        output,
+        containsAllInOrder(<Matcher>[
+          contains('Missing "implements: plugin_a" in "plugin" section.'),
+        ]),
```

Comment:
FYI I'm running into this problem with pigeon too where I assert the error text matches for a certain scenario but they are a bit fragile.  It would be better to assert on a codename for errors like some other compilers do.

`Error: Missing "implements: foo" in "plugin" section. [missing_implements]` then use `assert(error, contains('[missing_implements]')`

Reply:
I accidentally sent a PR from a branch that was on top of another PR, rather than master; this isn't actually part of the change. But in general this would apply to the whole tool though, so is worth considering.

The main difference relative to a compiler is that being directly read by humans is the only intended use case for these log messages. That means that the text itself is actually the thing I want to assert on, because if we somehow broke things such that instead of:
```
Missing "implements: plugin_a" in "plugin" section. [missing_implements]
```

it just printed:
```
[missing_implements]
```

that would be really bad.

I assume what you mean when you say they are fragile is that if the wording of the message changes, the tests need to change as well. Sometimes that is indeed mildly annoying (e.g., if it's a typo fix) but in most cases that's actually the desired outcome because having a clear, detailed error message is a critical part of the desired behavior.

(There's also no "ignore missing_implements" option, nor do we ever intend to build one since this isn't a general-purpose tool, which is the other main use case for having those kind of machine-friendly error names.)